### ISSUE TITLE
memtest program doesn't work for DE1-SOC

### ISSUE THREAD

**Issue by jackzhang1992** [MIPSfpga/mipsfpga-plus#25]
This time I use DE1-SOC board, 00_counter example works,
but when I downloaded the 04_memtest  "program.srec" to the board with UART-USB cable, nothing happens.

Do I need to change something in the verilog code, or main.c code?
In my opinion, there may be several reasons of my situation  as follows:
1) I didn't make the configuration right, like some to comment or uncomment something
2) timing constraint doesn't meet. Because I noticed some compilation summary that mentioned the timing constraint warnings for sdram pins. I am not sure if it is the main problem. 
3) the 04_memset program is not configured properly

For now, my configuration of the verilog code is just to uncomment   

**MFP_USE_SDRAM_MEMORY**

in  system_rtl/mfp_ahb_lite_matrix_config.vh
```
//
//  Memory type (choose one)
//
//`define MFP_USE_BYTE_MEMORY
//`define MFP_USE_WORD_MEMORY
//`define MFP_USE_BUSY_MEMORY
`define MFP_USE_SDRAM_MEMORY
```

thank you if anyone could give me some clue!

**Comment by jackzhang1992** [comment]
I add some code at the top of "boot.S"  to light the leds, it works! 
     
in 04_memtest/boot.S
```
start:	
	
	        lui   $8, 0xbf80     # $8 =  base I/O address (and LED address)
		lui  $10, 0xffff
		ori $10,0xffff
		sw    $10, 0($8)     # write  values to LEDs 
		addi $8,$0,0
		addi $10,$0,0
		

        mtc0    zero, C0_COUNT          # Clear cp0 Count (Used to measure boot time.)
       ...
```

but when I add some code in the main.c, LED doesn't change at all.

in 04_memtest\main.c
```
int main ()
{
	MFP_RED_LEDS   = 0xffff5555;
    ......
}
```

Conclusion is:
1>boot.S is loaded successfully to the reset ram area inside the FPGA chip,
2>main.S is not loaded  successfully to the SDRAM area outside of the FPGA chip.

But I am still working on this problem...
Thanks, FYI

**Comment by zhelnio** [comment]
@jackzhang1992 The problem is in SDRAM timings and clock shift. I have tested SDRAM memory interface module ( mfp_ahb_ram_sdram.v) only for DE10_lite. 
It also could work with other boards with SDRAM chip but only after some "hardware tuning":
- changing the PLL settings of SDRAM clock (it should be added on board code level if not exist);
- setting the correct values of SDRAM module params. These settings (DE10-Lite example):
https://github.com/MIPSfpga/mipsfpga-plus/blob/b7b0105e43925e5466b35c37fffb307f6ff9ee26/boards/de10_lite/de10_lite.v#L276-L288
https://github.com/MIPSfpga/mipsfpga-plus/blob/master/boards/de10_lite/board_config.vh

**Comment by jackzhang1992** [comment]
@jackzhang1992 

Thank you for this information. I will that later. 

I have figured that the SDRAM chip on DE1-SOC **IS42S16320D**,  is same as that on DE10-Lite.  So at first, I thought maybe I don't have to change the timing parameters. 

Any, I will give feedback when I have further results.

Thanks.

**Comment by zhelnio** [comment]
@jackzhang1992 the wire length on the PCB (FPGA <-> SDRAM) is different on these boards. So you may still have to do some tuning.

**Comment by jackzhang1992** [comment]
@zhelnio  That's true! Wire length are different. But are there some rules I can follow to tune the  these clk numbers. I haven't done this before, so I have no idea how to manage more than 10 parameters and make it run in the end.

But for now, I think I should have a look at the datasheet, right? 

Thanks.

```
//SDRAM parameters for 64MB (Terasic DE10-Lite)
`ifdef MFP_USE_SDRAM_MEMORY
     ......
    //values for fclk=50 MHz
    `define SDRAM_DELAY_nCKE        10000
    `define SDRAM_DELAY_tREF        360
    `define SDRAM_DELAY_tRP         0
    `define SDRAM_DELAY_tRFC        2
    `define SDRAM_DELAY_tMRD        0
    `define SDRAM_DELAY_tRCD        0
    `define SDRAM_DELAY_tCAS        0
    `define SDRAM_DELAY_afterREAD   0
    `define SDRAM_DELAY_afterWRITE  2
    `define SDRAM_COUNT_initAutoRef 8
`endif
```

**Comment by zhelnio** [comment]
@jackzhang1992 when I was adding an SDRAM support to DE10-Lite, I have created a special project for this task: https://github.com/zhelnio/ahb_lite_sdram 
It works successfully (at least for DE10-Lite), but also have some issues. I have just added their description to the project page: https://github.com/zhelnio/ahb_lite_sdram/issues
I have also wrote the detailed description of how this SDRAM module works. But this article is in Russian only (((
https://habrahabr.ru/post/321530/
https://habrahabr.ru/post/321532/

**Comment by zhelnio** [comment]
@jackzhang1992 another good link that can help you: https://github.com/zhelnio/ahb_lite_sdram/blob/master/doc/20170120_SDRAM_PLL_Tuning_en.pdf

**Comment by jackzhang1992** [comment]
@zhelnio  

I found a general way for the phase shift, quartus have a special PLL for DE series boards,   
ftp://ftp.altera.com/up/pub/Intel_Material/15.0/University_Program_IP_Cores/Clocks/Altera_UP_Clocks.pdf  .We could use this IP core in the right part of quartus main window.
**IP Catalog\Library\University Program\Clock\System and SDRAM Clocks for DE-series Boards** 
This helps us to solve the clock probelms.I use this PLL for  DRAM_CLK, but still the project doesn't work.

so now I think it was the problem of the second reason:

- setting the correct values of SDRAM module params. These settings (DE10-Lite example)

Now I plan to go backwards, and move further step by step. I  would try to make  https://github.com/zhelnio/ahb_lite_sdram this project to run first. Actutally I used the Signal Tap II, but still doesn't solve this problem, so I guess it wouldn't be that quick to solve a timing problem like this.

My next plan is to use icarus verilog+gtkwave to simulate and then learn how to tune the real hardware.

Thank you anyway!
I will post news here when I make progress in the future.
By the way, thank you for the Russian article, it would be great if I could  learn some Russian by solving this problem, hahaha!
